SIPS/sip-20.md

```diff
@@ -0,0 +1,155 @@
+        "dataType": "binary"
+      },
+      {
+        "type": "websocket",
+        "url": "wss://example.com/text/endpoint",
+        "dataType": "text"
+      },
+      {
+        "type": "blockchain",
+        "chainId": "eip155:1",
+        "events": [
+          "block"
+        ]
+      }
+    ]
+  }
+}
+```
+
+### Permission caveats
+
+Each caveat object MUST have `type` property that SHOULD be `websocket` or `blockchain`.
+If the `type` is `websocket`then the caveat MUST have `url` and `dataType` properties.
+The `url` MUST be string.
+The `dataType` MUST be string that MUST be `"binary"` or `"text"`.
+The `chainId` MUST be string that represents supported blockchain.
+The `events` SHOULD be an array of strings which represent name of the events for subscription.
+
+- `type` property represents the type of data source.
+- `url` is the WebSocket URL.
+- `dataType` is type of data that WebSocket returns.
+- `chainId` is the blockchain ID.
+- `events` is the array of events to subscribe.
+
+The caveats MUST NOT have duplicate objects with the same `url` or `chainId` properties.
+
+The `url` MUST start with `wss://` which is a protocol indicator for secure WebSocket connection.
+
+### Snap implementation (request handler)
+
+This SIP proposes the following RPC request handler to execute implemented functionality for external data events:
+
+#### OnExternalDataHandler
+
+Example:
+```typescript
+import { OnExternalDataHandler } from '@metamask/snaps-types';
+import { assert } from '@metamask/utils';
+
+export const onExternalData: OnExternalDataHandler = ({ id, type, data, source }) => {
+  assert(type === 'websocket'); // `data` is inferred as `WebSocketData`.
+  assert(data.type === 'text'); // `message` is inferred as `string`.
+
+  // Now the Snap can do whatever is needed with the message.
+  const json = JSON.parse(data.message);
+};
+```
+
+Snap that wants to use external data sources MUST have `onExternalData` function export.
+
+Exported function SHOULD accept an object parameter that MAY have `id`, `type`, `data` and `source` properties.
+
+This export SHOULD be called every time the WebSocket client or Blockchain event listener receives some data or event.
+
+The parameters represent the following:
+- `id`: Unique ID of an event generated internally on the controller side. This can be used for tracking each event
+  in case of storage needs or debugging. The `id` MUST be unique string.
+- `data`: SHOULD be the raw message that the WebSocket client received. Depending on the `type`, this SHOULD be
+  either a `string` or an `ArrayBuffer`.
+- `type`: The type of the message that was sent. This SHOULD be either `text` or `binary`.
+- `source`: The URL that the message was sent from. Snaps can use this to differentiate between different endpoints.
+  This SHOULD be the exact same URL as specified in manifest.
+
+The RECOMMENDED specification of types of `OnExternalDataHandler` and its parameters:
```

Comment:
I don't think types should be a recommendation?

Reply:
Removed.